include/binaryfusefilter.h

```diff
@@ -858,4 +860,91 @@ static inline bool binary_fuse8_deserialize(binary_fuse8_t * filter, const char
   return true;
 }
 
+// minimal bitfield implementation
+typedef uint8_t bitf_t;
+#define bitf_w (sizeof(bitf_t) * 8)
+#define bitf_sz(bits) (((bits) + bitf_w - 1) / bitf_w)
+#define bitf_word(bit) (bit / bitf_w)
+#define bitf_bit(bit) ((1U << (bit % bitf_w)) % 256)
+
+#define ser(buf, lim, src) do {			\
+	if ((buf) + sizeof src > (lim))		\
+	  return (0);				\
+	memcpy(buf, &src, sizeof src);		\
+	buf += sizeof src;			\
+} while (0)
+
+#define deser(dst, buf, lim) do {		\
+	if ((buf) + sizeof dst > (lim))		\
+	  return (false);			\
+	memcpy(&dst, buf, sizeof dst);		\
+	buf += sizeof dst;			\
+} while (0)
+
+// return required space for binary_fuse8_pack()
+static inline size_t binary_fuse8_pack_bytes(const binary_fuse8_t *filter)
+{
+  size_t sz = 0;
+  sz += sizeof filter->Seed;
+  sz += sizeof filter->Size;
+  sz += bitf_sz(filter->ArrayLength);
+  for (size_t i = 0; i < filter->ArrayLength; i++) {
+    if (filter->Fingerprints[i] == 0)
+      continue;
+    sz += sizeof filter->Fingerprints[i];
+  }
+  return (sz);
+}
+
+// serialize as packed format, return size used or 0 for insufficient space
+static inline size_t binary_fuse8_pack(const binary_fuse8_t *filter, uint8_t *buffer, size_t space) {
+  uint8_t *buf = buffer;
+  uint8_t *e = buf + space;
+
+  ser(buf, e, filter->Seed);
+  ser(buf, e, filter->Size);
+  size_t bsz = bitf_sz(filter->ArrayLength);
+  if (buf + bsz > e)
+    return (false);
+  // XXX only works because of byte-aligned bitf type
+  bitf_t *b = (bitf_t *)(void *)buf;
+  buf += bsz;
+  memset(b, 0, bsz);
+  
+  for (size_t i = 0; i < filter->ArrayLength; i++) {
+    if (filter->Fingerprints[i] == 0)
+      continue;
+    b[bitf_word(i)] |= bitf_bit(i);
+    ser(buf, e, filter->Fingerprints[i]);
+  }
+  assert(buf > buffer);
```

Comment:
Instead of an assert, which gets compiled away when compiling with NDEBUG, please consider returning an error code. Typically, one returns (size_t)-1.

Reply:
Generally, I do not use assert and many systems will disallow it. An assert compiles to an abort/terminate and/or it gets compiled away.

In this instance, it appears that you want to check for buffer overruns... I would much rather than we prove that it cannot happen rather than use an assert.

Reply:
The assert was not for an error condition, it was to ensure that the difference between the end and start pointer is positive. I am going to remove it.